File: monads/Maybe.js

john = {
    firstName: "John",
    middleName: "Edward",
    lastName: "Doe",
    homeAddress: {
        line1: "123 W. Somewhere Dr.",
        line2: "Apt. 334",
        city: "Kansas City",
        state: "Missouri",
        zipCode: 64108
    },
    workAddress: {
        line1: "432 N. Nowhere Dr.",
        line2: "Ste. 643",
        city: "Kansas City",
        state: "Missouri",
        zipCode: 64109
    }
};

jane = {
    firstName: "Jane",
    lastName: "Doe"
}

console.log(john.middleName);
console.log(jane.middleName);

//console.log(john.homeAddress.line2);
//console.log(jane.homeAddress.line2);

function getHomeLine2(obj) {
    var a = obj.homeAddress;
    if (a) return a.line2;
    else return null;
}

function getWorkLine2(obj) {
    var a = obj.workAddress;
    if (a) return a.line2;
    else return null;
}

console.log(getHomeLine2(john));
console.log(getHomeLine2(jane));

console.log(getWorkLine2(john));
console.log(getWorkLine2(jane));

File: monads/Maybe.hs
import           Prelude hiding (Maybe (..))

data Maybe a = Nothing
             | Just a
             deriving (Eq, Show)

data Person = Person
            { firstName   :: String
            , middleName  :: Maybe String
            , lastName    :: String
            , homeAddress :: Maybe Address
            , workAddress :: Maybe Address
            } deriving (Eq, Show)

data Address = Address
            { line1   :: String
            , line2   :: Maybe String
            , city    :: String
            , state   :: String
            , zipCode :: Int
            } deriving (Eq, Show)

john :: Person
john = Person "John" (Just "Edward") "Doe"
        (Just (Address "123 W. Somewhere Dr." (Just "Apt. 334") "Kansas City" "Missouri" 64108))
        (Just (Address "432 N. Nowhere Dr." (Just "Ste. 43") "Kansas City" "Missouri" 64109))

jane :: Person
jane = Person "Jane" Nothing "Doe" Nothing Nothing


--  Non-Monadic Solution
getHomeLine2 :: Person -> Maybe String
getHomeLine2 p = case homeAddress p of
                    Nothing   -> Nothing
                    Just addr -> line2 addr

getWorkLine2 :: Person -> Maybe String
getWorkLine2 p = case workAddress p of
                    Nothing   -> Nothing
                    Just addr -> line2 addr

--  Monadic Solution
instance Monad Maybe where
    (Just x) >>= k      = k x
    Nothing  >>= _      = Nothing

    return              = Just

main :: IO ()
main = do
        print $ middleName john
        print $ middleName jane

        --print $ line2 $ homeAddress john
        --print $ line2 $ homeAddress jane

--  Non-Monadic Solution
        print $ getHomeLine2 john
        print $ getHomeLine2 jane

        print $ getWorkLine2 john
        print $ getWorkLine2 jane

--  Monadic Solution
        print $ homeAddress john >>= line2
        print $ homeAddress jane >>= line2

        print $ workAddress john >>= line2
        print $ workAddress jane >>= line2
